hal/src/dma.rs

```diff
@@ -0,0 +1,1075 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+mod controller;
+mod status;
+
+#[cfg(not(feature = "samd51"))]
+use self::controller::samd21::{IRQn, IRQN};
+#[cfg(feature = "samd51")]
+use self::controller::samd51::{ChannelRegisters, IRQn};
+use self::controller::*;
+pub use self::status::Status;
+#[cfg(feature = "samd51")]
+use crate::target_device::dmac::chctrla::{TRIGACTW, TRIGSRCW};
+#[cfg(not(feature = "samd51"))]
+use crate::target_device::dmac::chctrlb::{TRIGACTW, TRIGSRCW};
+use crate::target_device::DMAC;
+use core::{
+    mem,
+    sync::atomic::{self, Ordering},
+};
+use vcell::VolatileCell;
+
+/// Maximum number of jobs to resume
+const MAX_JOB_RESUME_COUNT: usize = 10000;
+
+/// DMA clock source
+#[cfg(feature = "samd51")]
+type ClockSource = crate::target_device::MCLK;
+
+/// DMA clock source
+#[cfg(not(feature = "samd51"))]
+type ClockSource = crate::target_device::PM;
+
+/// DMA callback types
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum CallbackType {
+    /// Transfer errors are flagged if a bus error is detected during an AHB
+    /// access or when he DMAC fetches an invalid descriptor.
+    TransferError,
+
+    /// Transfer finished successfully
+    TransferDone,
+
+    /// Channel suspended
+    ChannelSuspend,
+}
+
+/// DMA callbacks
+#[derive(Copy, Clone, Default)]
+pub struct Callbacks {
+    /// Transfer error
+    pub transfer_error: Option<fn(&mut DMA)>,
+
+    /// Transfer done
+    pub transfer_done: Option<fn(&mut DMA)>,
+
+    /// Channel suspended
+    pub channel_suspend: Option<fn(&mut DMA)>,
+}
+
+/// Direct Memory Access
+pub struct DMA<'desc> {
+    /// DMA controller
+    dmac: DMAC,
+
+    /// Channel
+    channel: Option<ChannelId>,
+
+    /// Job status
+    job_status: VolatileCell<Status>,
+
+    /// Has descriptors?
+    has_descriptors: bool,
+
+    /// Loop flag
+    loop_flag: bool,
+
+    /// Peripheral trigger
+    peripheral_trigger: TRIGSRCW,
+
+    /// Trigger action
+    trigger_action: TRIGACTW,
+
+    /// Callbacks
+    callbacks: Callbacks,
+
+    /// Descriptor list
+    descriptor_list: &'desc DescriptorList,
+}
+
+pub struct InterruptState {
```

Comment:
This was all (somewhat necessarily) global state in ZeroDMA. Any ideas what the Rust code should do? I shoved it all in a struct for now, but it then becomes a parameter to every method that needs to acquire a critical section.

Reply:
The need for this was eliminated by switching to `cortex_m::interrupt::Mutex`